doc/Precision_Guidelines_for_Ngen_Code_Validation.md

```diff
@@ -0,0 +1,36 @@
+
+
+**Possible reasons for precision discrepancies between CFE and BMI-CFE:**
+
+The double precision numbers have an accuracy of 15 decimal digits, the loss of accuracy from rounding off error is expected, but precisely how many digit loss probably varies and depends on the specific code. Since BMI-CFE code has additional interface and difference in coding details, we hypothesize the memory allocation, cache usage, data sets loading into the registers may be different. These likely affect the rounding off processes and cause the least significant digits to differ. According to some literature, computer hardware architecture, compiler, even optimization may all affect numerical precision. 
+        
+In our CFE and BMI-CFE validation example, one particular case, deserves attention, which is: a key input variable, the rain rate, or specifically: precip_kg_per_m2, is a float point number, this could potentially reduces the precision in comparison between the two models even though all other variables are double, in consideration of rounding-off error in machine precision. There are many excellent discussions on rounding-off errors in literature due to machine precision, we refer interested readers to literature<sup>1-4</sup> for details. 
+
+On the other hand, the BMI-CFE and the original CFE demonstrate remarkable stability with 8-9 decimal point accuracy even after 720 time steps (hours). What are the possible reasons? We hypothesize two possible causes:
+1. The codes do not use space grids, all flow values are single valued for the catchment, i.e., there is no instability introduced by spatial matrices such as those in solving finite difference equations.
+2. Mass balance: In the evolution in time, the flows are constrained by mass balance. The mass balance, like energy conservation in classical molecular dynamics, should constrain the flow variation to a certain range, thus limiting the error propagation magnification in time (possibly through error cancellation).
+
+
+References and Links  
+1. [What Every Computer Scientist Should Know About Floating-Point Arithmetic](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html).

```

Comment:
I  like this reference, I go back to it often!

Reply:
The same for me. though I probably haven't read as thoroughly as you did.